godot-core/src/builtin/quaternion.rs

```diff
@@ -0,0 +1,288 @@
+            -sin_a1 * sin_a2 * cos_a3 + cos_a1 * cos_a2 * sin_a3,
+            sin_a1 * sin_a2 * sin_a3 + cos_a1 * cos_a2 * cos_a3,
+        )
+    }
+
+    pub fn get_angle(self) -> f32 {
+        2.0 * self.w.acos()
+    }
+
+    pub fn get_axis(self) -> Vector3 {
+        if self.w.abs() > 1.0 - CMP_EPSILON {
+            return Vector3::new(self.x, self.y, self.z);
+        }
+        let r = 1.0 / (1.0 - self.w * self.w).sqrt();
+        Vector3::new(self.x * r, self.y * r, self.z * r)
+    }
+
+    pub fn inverse(self) -> Self {
+        Quaternion::new(-self.x, -self.y, -self.z, self.w)
+    }
+
+    pub fn is_equal_approx(self, to: Self) -> bool {
+        is_equal_approx(self.x, to.x)
+            && is_equal_approx(self.y, to.y)
+            && is_equal_approx(self.z, to.z)
+            && is_equal_approx(self.w, to.w)
+    }
+
+    pub fn is_finite(self) -> bool {
+        self.x.is_finite() && self.y.is_finite() && self.z.is_finite() && self.w.is_finite()
+    }
+
+    pub fn is_normalized(self) -> bool {
+        is_equal_approx(self.length_squared(), 1.0) /*,UNIT_EPSILON)*/
+    }
+
+    pub fn length(self) -> f32 {
+        self.length_squared().sqrt()
+    }
+
+    pub fn length_squared(self) -> f32 {
+        self.dot(self)
+    }
+
+    pub fn log(self) -> Self {
+        let v = self.get_axis() * self.get_angle();
+        Self::new(v.x, v.y, v.z, 0.0)
+    }
+
+    pub fn normalized(self) -> Self {
+        self / self.length()
+    }
+
+    pub fn slerp(self, to: Self, weight: f32) -> Self {
+        let mut cosom = self.dot(to);
+        let to1: Self;
+        let omega: f32;
+        let sinom: f32;
+        let scale0: f32;
+        let scale1: f32;
+        if cosom < 0.0 {
+            cosom = -cosom;
+            to1 = -to;
+        } else {
+            to1 = to;
+        }
+
+        if 1.0 - cosom > CMP_EPSILON {
+            omega = cosom.acos();
+            sinom = omega.sin();
+            scale0 = ((1.0 - weight) * omega).sin() / sinom;
+            scale1 = (weight * omega).sin() / sinom;
+        } else {
+            scale0 = 1.0 - weight;
+            scale1 = weight;
+        }
+        Self::new(
+            scale0 * self.x + scale1 * to1.x,
+            scale0 * self.y + scale1 * to1.y,
+            scale0 * self.z + scale1 * to1.z,
+            scale0 * self.w + scale1 * to1.w,
+        )
+    }
+
+    pub fn slerpni(self, to: Self, weight: f32) -> Self {
+        let dot = self.dot(to);
+        if dot.abs() > 0.9999 {
+            return self;
+        }
+        let theta = dot.acos();
+        let sin_t = 1.0 / theta.sin();
+        let new_factor = (weight * theta).sin() * sin_t;
+        let inv_factor = ((1.0 - weight) * theta).sin() * sin_t;
+        Self::new(
+            inv_factor * self.x + new_factor * to.x,
+            inv_factor * self.y + new_factor * to.y,
+            inv_factor * self.z + new_factor * to.z,
+            inv_factor * self.w + new_factor * to.w,
+        )
```

Comment:
This could reuse the operators, no?
```rs
inv_factor * self + new_factor * to
```

Reply:
this is true. must be me being so used to C at this point that i instinctively construct a new one